backupccl: add dry_run option to RESTORE

msbutler opened the pull request:
backupccl: add dry_run option to RESTORE
    
Previously, users could not test whether their RESTORE would run correctly
without actually running a RESTORE. The dry_run option allows users to test
whether their table, database, or cluster level RESTORE will work on a test
cluster. A 'long' dry run RESTORE will perform all aspects of the restore,
short of writing actual data to disk. A 'short' dry run RESTORE will plan the
restore and check that all needed SSTs exist. At the end of the dry run
restore, all state in from the test cluster will be recovered.

Fixes: #67935

Release note (sql change): user can now test whether a RESTORE will run
successfully by using the dry_run option, in a test cluster. dry_run='short'
plans the restore and checks that all needed backup files exist and
dry_run='long' conducts the whole restore, short of writing data to disk.

Release justification: low risk, high benefit changes to existing functionality

msbutler committed:
backupccl: add dry_run option to RESTORE

Previously, users could not test whether their RESTORE would run correctly
without actually running a RESTORE. The dry_run option allows users to test
whether their table, database, or cluster level RESTORE will work on a test
cluster. A 'long' dry run RESTORE will perform all aspects of the restore,
short of writing actual data to disk. A 'short' dry run RESTORE will plan the
restore and check that all needed SSTs exist. At the end of the dry run
restore, all state in from the test cluster will be recovered.

Fixes: #67935

Release note (sql change): user can now test whether a RESTORE will run
successfully by using the dry_run option, in a test cluster. dry_run='short'
plans the restore and checks that all needed backup files exist and
dry_run='long' conducts the whole restore, short of writing data to disk.

Release justification: low risk, high benefit changes to existing functionality

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/74474)
<!-- Reviewable:end -->

shermanCRL commented:
Nice. Looks like the trick here is tracking down anything that mutates -- I wonder if it might be fragile in the future, when we change RESTORE logic but don’t catch the dry_run case. 

Is there a high-level test that can verify “nothing has changed” -- i.e. no tables deleted, maybe same # of rows, etc?

msbutler commented:
Yeah, not only am I worried about future changes, I'm also not sure if I missed any strange corner cases where some RESTORE option triggers an early write to disk that I haven't taken into account.  I'm not sure how to account for this, other than by adding a dry run restore to a test on each restore option... which doesn't seem very clean.

miretskiy reviewed (commented):
Let's not track anything down -- can we inject a cloud storage factory that returns cloud storage which disallows writes?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74474#-:-Msg65GXA2QfGFi6b_SA:b-g9jac4)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dt)



<!-- Sent from Reviewable.io -->

msbutler commented:
@miretskiy I'm less worried about backup data from cloud storage corrupting the target cluster because I added the `noopSender` which will noop all AddSSTable requests called by a `dry_run` restore. 

https://github.com/cockroachdb/cockroach/blob/a784165c52746aef0f748620025df6ab55db43e5/pkg/ccl/backupccl/restore_data_processor.go#L369-L385


I'm more worried about sneaky spots in the restore codebase that modify system tables or existing data in the cluster before or after the restore processors have done their work. E.g. in planning ,we drop the following dbs during a full cluster restore:

https://github.com/cockroachdb/cockroach/blob/a784165c52746aef0f748620025df6ab55db43e5/pkg/ccl/backupccl/restore_planning.go#L2032-L2040

I wonder if we can leverage the code used to clean up a target cluster after restore fails during/after execution?

miretskiy commented:
@msbutler Good point; my suggestion still stands though.  I guess it might require some interfaces added (*e.g. for dropping descriptors, etc) that you then replace with no-op implementations.

msbutler commented:
a potentially safer idea that yev suggested: leverage grant permissions to make the restore read only. I'll need to think a bit more about this idea.

adityamaru commented:
> I wonder if we can leverage the code used to clean up a target cluster after restore fails during/after execution?

I think this is worth exploring a little more, it feels the most sound to me. `OnFailOrCancel` is already responsible for leaving things as they were if the job fails. When we have found bugs where this contract is violated, we modify this hook to fix them, so there is a high degree of confidence in its correctness.

I haven't looked at the code very closely but I'm wary of `if DryRun` checks scattered throughout the restore code, this is bound to go stale as the code evolves. Can we do something where `noop` AddSSTables + a call to `OnFailOrCancel` gets us a "dry run"? There are probably some things like dropping and recreating `defaultdb` and `postgres` during dry run that we can avoid, but those can be special-cased.

msbutler commented on pkg/ccl/backupccl/restore_job.go:
```diff
@@ -1237,7 +1238,9 @@ func createImportingDescriptors(
 	}
 
-	if !details.PrepareCompleted {
+	if !details.PrepareCompleted && !details.DryRun {
```
by skipping this branch, I skip switching the status to JobRunning. I need to fix this.

msbutler commented:
In the latest commit, I prototyped a wrapper around all objects used in `restore_job.go` to manipulate data:

```
type restoreExec interface{
	NonDryJobExecCtx() sql.JobExecContext
	NonDryExecCfg() *sql.ExecutorConfig
	SafeExecCfg() *sql.ExecutorConfig
	SafeDBHandler() safeDB
	SafeDescTxn(context.Context, func(ctx context.Context, txn *kv.Txn,
		descsCol *descs.Collection) error) error
}
```
 The goal of `restoreExec` is to reduce the amount of special casing needed to implement dry run and to prevent future changes from screwing up dry run.

Limits of this implementation:
- these guardrails are quite flimsy. It is still the responsibility of the programmer to call the right object to ensure a safe dry run restore. 
- i added a lot of scaffolding and deleted only two `if dry_run` branches.
- as they stand now, the wrappers didn't significantly increase the dry run code coverage, though the next steps below could.

Next Steps, if folks think this approach deserves more thought:
- Currently any function wrapped in `sql.DescTxn()` or `db.Txn()` that also sends writes to kv does _not_ get called in a dry run, significantly limiting code coverage. I'll think about how to call these functions safely, though I'm not sure this is possible without manipulating code outside bulk's purview.
- implement a similar wrapper for `restore_planning.go`
- add a unit test that fails if `OnFailOrCancel` conducted significant clean up work on a dry run. As mentioned before, `OnFailOrCancel` should be treated as a last ditch effort to preserve cluster state. Ideally no dry run restore on master should depend on `OnFailOrCancel`
- Cluster Level Dry Run Restore on an active cluster.  In the spirit of `RESTORE DATABASE .... WITH new_db_name` I could rename every descriptor from the cluster backup, avoiding all descriptor collisions. I'm pretty confident this will work.

msbutler commented:
big update: no need to review my latest change until Liv learns from customers whether they would prefer running dry run restore on a production or test cluster. If run on a test cluster, we'll need to verify if customers are ok with system tables getting modified by dry run. If so, this implementation becomes a lot easier.

adityamaru commented:
>if customers are ok with system tables getting modified by dry run

Is this wrt to the rows we insert into the system tables as part of a cluster restore? I assume a dry run could just not do that whole step, where it inserts rows into the restoring cluster system tables? It seems to me that the main value proposition of dry run restore would be to restore user data.

msbutler commented:
i was primarily referring to modifying the descriptors table in a table, db, or cluster level restore. The upshot is, if a customer is ok with dry run (temporarily) modifying on disk data, then we don't need all this special casing for a dry run restore. Rather, we can run a regular old restore, short of issuing split and scatter requests and writing SSTs to disk.

msbutler commented:
Just pushed another  draft with some substantial changes. Dry Run Restore is now designed to run on a test cluster, which gives us significantly more code coverage than the previous approach. It basically does everything— it writes restoring descriptors to disk, uses the conventional split and scatter processor, deletes the user and default db table. After the dry run completes, it calls OnFailOrCancel to revert the cluster back to its pre dry run restore state. 

The only two differences compared to a regular restore are: 1) it still doesn’t write user data to disk in the restore processor; 2) in a cluster dry run, it doesn’t restore the system tables.

I have a couple open design questions which I hope to answer before I write new tests and implement the ‘fast’ and ‘slow’ dry run variants.

- A cluster level dry run restore currently fails during clean up (i.e. in OnFailOrCancel ) if it [calls](https://github.com/cockroachdb/cockroach/blob/master/pkg/ccl/backupccl/restore_job.go#L1457) `restoreSystemTables` before cleanup. This actually seems like a vulnerability for a regular cluster restore— if a restore fails after `restoreSystemTables` , I think we’d then fail in `OnFailOrCancel` , potentially leaving the cluster in an inconsistent state. Thoughts?
- since we don’t care about temporarily modifying on disk data, is it alright to use the normal split and scatter processor?
- to ensure customers only run dry run on a test cluster, should i add another cluster setting called `enable_dry_run_restore` that defaults to false? What other safeguards should I add to this?

msbutler commented:
ack sry, i screwed something up. hold on. pushing a fix.

dt commented:
> if a restore fails after restoreSystemTables , I think we’d then fail in OnFailOrCancel

Hm. This sorta makes sense: `restoreSystemTables` is destructive, right? we delete data out of the system table and add in the data from the restored version of the table and then commit that txn, at which point we've deleted that data; it could be GC'ed and then there's no undoing that. It seems like we should just be focusing on making sure that that is the very last thing we do s.t. we don't fail after. I wonder if non-fixed system table IDs help us here, in that maybe we could just swap the old table out for the new one without deleting anything.

dt commented:
> since we don’t care about temporarily modifying on disk data, is it alright to use the normal split and scatter processor?

Eh. The less we mutate, the better. Indeed, I think we should _aim_ to have a no-writes-outside-jobs dry run eventually even if for now we have, say, some temporary staged descriptors that come and go. Also, splits are slow.

msbutler commented:
closing, as https://github.com/cockroachdb/cockroach/pull/85231 implemented a restore validation alternative.
